### DISCUSSION TITLE
Is there a Limit to parallel Requests?

### DISCUSSION THREAD

**Discussion by EricKrg** [appwrite/appwrite#4870] [Q&A]
I am not sure if i missed something in the docs or if this is maybe actual unexpected behavior.

I implemented a function which deletes a document, this document also holds a List of IDs as reference to other documents as well as File-IDs to uploads, these dependent Docs and Uploads also need to be deleted when the Parent Doc is deleted. The idea was to collect all these request and to send them all at once with `Promise.all`.  My first approach was doing it like this: 

```ts
[...]
// interfaces shown here have the same representation in the appwrite backend
export interface TaskRecord {
  id: string
  task: string
  taskState: boolean
  attachments: string[] // list with IDs of dependent uploads (Files) and documents
  updated: string
}

export interface Attachment {
  filename: string
  file_id: string
}
[...]
async deleteTask (taskRecord: TaskRecord): Promise<boolean> {
    const deleteDoc = this.database.deleteDocument(Server.databaseID, Server.collectionID, taskRecord.id)
    const calls: Array<Promise<any>> = [deleteDoc]
    // delete all attachments
    taskRecord.attachments.forEach(i => {
       calls.push(this.storage.deleteFile(Server.bucketID, i))
       calls.push(this.database.deleteDocument(Server.databaseID, Server.attachmentID, i))
    })
   
    try {
      const res = await Promise.all(calls)
      return true
    } catch (error) {
      console.warn(`Something went wrong deleting the Task with the id ${taskRecord.id}`, error)
      return false
    }
  }
```

But doing it that way gives me a rather vague backend error when deleting 2 or more dependent docs/uploads:

Response: 
```
{"message":"Server Error","code":500,"type":"general_unknown","version":"1.1.0"}
```
Response-Header:
```
HTTP/1.1 500 Internal Server Error
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Origin, Cookie, Set-Cookie, X-Requested-With, Content-Type, Access-Control-Allow-Origin, Access-Control-Request-Headers, Accept, X-Appwrite-Project, X-Appwrite-Key, X-Appwrite-Locale, X-Appwrite-Mode, X-Appwrite-JWT, X-Appwrite-Response-Format, X-SDK-Version, X-SDK-Name, X-SDK-Language, X-SDK-Platform, X-Appwrite-ID, Content-Range, Range, Cache-Control, Expires, Pragma
Access-Control-Allow-Methods: GET, POST, PUT, PATCH, DELETE
Access-Control-Allow-Origin: http://localhost:3000
Access-Control-Expose-Headers: X-Fallback-Cookies
Cache-Control: no-cache, no-store, must-revalidate
Content-Encoding: br
Content-Length: 70
Content-Type: application/json; charset=UTF-8
Date: Fri, 16 Dec 2022 14:00:57 GMT
Expires: 0
Pragma: no-cache
Server: Appwrite
X-Content-Type-Options: nosniff
X-Debug-Fallback: false
X-Debug-Speed: 0.10433292388916
X-Ratelimit-Limit: 60
X-Ratelimit-Remaining: 59
X-Ratelimit-Reset: 1671199260
```
An Intresting observation i made is that the X-Ratelimit-Remaining: 59 does not change between the parallel started requests, so maybe thats connected to the problem i am facing. 

I only got it to work with the following workaround: 
```ts
async deleteTask (taskRecord: TaskRecord): Promise<boolean> {
    const deleteDoc = this.database.deleteDocument(Server.databaseID, Server.collectionID, taskRecord.id!)
    const calls: Array<Promise<any>> = [deleteDoc]
    const sleep = async (ms: number): Promise<void> => await new Promise(resolve => setTimeout(resolve, ms))
    try {
      const res = await Promise.all(calls)
      for (const attachment of taskRecord.attachments) {
        await sleep(100)
        console.log('delete doc')
        await this.database.deleteDocument(Server.databaseID, Server.attachmentID, attachment)
        await sleep(100)
        console.log('delete file')
        await this.storage.deleteFile(Server.bucketID, attachment)
      }
      console.log(res)
      return true
    } catch (error) {
      console.warn(`Something went wrong deleting the Task with the id ${taskRecord.id}`, error)
      return false
    }
  }
```
Which is not really what i want to do for several reasons. 

I know that i could write a server function for this problem, but i think that should also work on the client side?
If you need more Information i am happy to share them.

**Comment by stnguyen90** [comment]
> {"message":"Server Error","code":500,"type":"general_unknown","version":"1.1.0"}

There should be a log message in your appwrite container docker logs for the 500 error. Would you please check that and provide?

> An Intresting observation i made is that the X-Ratelimit-Remaining: 59 does not change between the parallel started requests, so maybe thats connected to the problem i am facing.

There is a slight problem with the rate limits at the moment where they aren't counted properly for requests in a short timespan. This could also be the reason for the 500 error. 

> I know that i could write a server function for this problem, but i think that should also work on the client side?

You really might want to move this server-side because of rate limits.

**Reply by EricKrg** [reply]
@stnguyen90 I checked the logs for the appwrite container, which were the following:

```
[Error] Timestamp: 2022-12-22T09:21:15+00:00
[Error] Method: DELETE
[Error] URL: /v1/storage/buckets/:bucketId/files/:fileId
[Error] Type: Utopia\Database\Exception\Duplicate
[Error] Message: Duplicated document: Duplicate entry 'ip:10.0.2.2,method:DELETE,url:localhost/v1/storage/buckets/:b...' for key 'unique1'
[Error] File: /usr/src/code/vendor/utopia-php/database/src/Database/Adapter/MariaDB.php
[Error] Line: 547
[Error] Timestamp: 2022-12-22T09:21:15+00:00
[Error] Method: DELETE
[Error] URL: /v1/databases/:databaseId/collections/:collectionId/documents/:documentId
[Error] Type: Utopia\Database\Exception\Duplicate
[Error] Message: Duplicated document: Duplicate entry 'ip:10.0.2.2,method:DELETE,url:localhost/v1/databases/:databas...' for key 'unique1'
[Error] File: /usr/src/code/vendor/utopia-php/database/src/Database/Adapter/MariaDB.php
[Error] Line: 547
```
I tested different scenarios, but it usually failed to delete the attachment files and docs if more then 3 Attachments were present in the parent document. 

I only found a corresponding error log in the appwrite container, if you need more just tell me what you need.  If necessary i can also provide the complete project setup.

**Reply by stnguyen90** [reply]
Ya, this looks related to the rate limit. See https://github.com/appwrite/appwrite/issues/4642. For now, you might need to [disable rate limits](https://appwrite.io/docs/rate-limits#disable-limits) while we get the bug fixed.
